File: 04_08_First_Common_Ancestor.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode lowestCommonAncestor(TreeNode current, TreeNode p, TreeNode q) {
        if (current == null) {
            return null;
        }
        
        if (current == p || current == q) {
            return current;
        }
        
        TreeNode left = lowestCommonAncestor(current.left, p, q);
        
        TreeNode right = lowestCommonAncestor(current.right, p, q);
        
        if (left != null && right != null) {
            return current;
        }
        
        if (left != null) {
            return left;
        }
        
        if (right != null) {
            return right;
        }
        
        return null;
    }
}

File: 04_08_First_Common_Ancestor.kt
/**
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int = 0) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */

class Solution {
    fun lowestCommonAncestor(root: TreeNode?, p: TreeNode?, q: TreeNode?): TreeNode? {
        
        if (root == null) return null
        
        if (root == p || root == q) return root
        
        val left = lowestCommonAncestor(root.left, p, q)
        val right = lowestCommonAncestor(root.right, p, q)
        
        if (left != null && right != null) return root
        
        return left ?: right
    }
}